### ISSUE TITLE
[Feature Request] Warning during optimization

### ISSUE THREAD

**Issue by aleczhanshi** [pytorch/botorch#401]
Hi there,

I've come across the warning below a few times now and then. Could anyone explain the causes and if I should ignore it? Thanks!

```python
UserWarning: Runtime Error when computing Cholesky decomposition: cholesky_cpu: For batch 0: U(5,5) is zero, singular U.. Using RootDecomposition.
```

**Comment by saitcakmak** [comment]
As you know, the calculations with GPs involve a lot of multivariate normal random variables. Let T ~ Normal (\mu, \Sigma) be the multivariate random variable of interest. With Z as the standard normal random variable of appropriate dimension, we can write T = \mu + \Sigma^{1/2} Z (equality in distribution). \Sigma^{1/2} is some matrix satisfying \Sigma = \Sigma^{1/2} \Sigma^{1/2 transpose}. A particular choice of \Sigma^{1/2} is the Chelosky decomposition of \Sigma. This is the default option in BoTorch and has some numerical advantages over other alternatives. For Chelosky decomposition to exist, we need a positive definite \Sigma. You get the above error when you have 0 in the diagonal of \Sigma (it is not positive definite). I believe this happens due to some numerical rounding when you have lots of samples, samples clustered around each other, and/or low noise observations. BoTorch typically gets around this by using RootDecomposition instead, which gives you an alternative \Sigma^{1/2}. I personally haven't seen any harm from ignoring it so far. My understanding is that things still work as they should, only a bit less efficiently. Hope this helps.

You can check this paper for the reparametrization trick in case you're not familiar with it: https://arxiv.org/abs/1712.00424

**Comment by Balandat** [comment]
Indeed, this may happen if a covariance matrix is near singular. By default, BoTorch (via GPyTorch) will add some small positive value (jitter) to the diagonal and retry, which typically works (though unfortunately there is a huge perf issue in pytorch where the error we catch takes forever to surface, which can slow things down significantly, see https://github.com/pytorch/pytorch/issues/34272). 

@aleczhanshi, I haven't seen this particular warning in a while, which is coming from GPyTorch, and should only be used if the repeated addition of jitter does not solve the problem. I'm curious in what circumstances this is happening, would you have a minimal reproducible example?

**Comment by saitcakmak** [comment]
@Balandat I observe this error regularly. Here is a minimal example. All you need is to take an excessive amount of samples from GP.
```
import torch
from botorch.models import SingleTaskGP
from botorch.fit import fit_gpytorch_model
from gpytorch.mlls import ExactMarginalLogLikelihood
from botorch.models.transforms import Standardize
from botorch.test_functions import Branin

torch.manual_seed(0)
function = Branin(noise_std=0.1, negate=True)
d = function.dim  # dimension of train_X
n = 50  # training samples

train_X = torch.rand((n, d))
train_Y = function(train_X).unsqueeze(-1)

gp = SingleTaskGP(train_X, train_Y, likelihood=None, outcome_transform=Standardize(m=1))
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_model(mll)

test_X = torch.rand((500, d))
post = gp.posterior(test_X)
samples = post.rsample(torch.Size([100, 100]))
```

**Comment by Balandat** [comment]
Yeah so this happens b/c when sampling from the posterior we need to perform the decomposition of the posterior covariance `K(test_X, test_X)` (without adding noise, as is done during the inference step when you consider `K(train_X, train_X) + sigma^2 I`). 

The issue is not the number of samples, but the number of points at which you draw the samples. With 500 points is a low-dimensional space the likelihood of two points being very close together is quite high. But if you compute the covariance `K(test_X, test_X)` where `test_X` has repeated points, the resulting matrix is necessarily singular. If you're close to repeated, you're close to singular (close enough for numerical precision to be an issue - this will likely work much better in `double` dtype, btw.). 

If you don't care much about the loss of precision, it's fine to add a small jitter value anyway. If you do care about doing proper sampling, one could use a modified Cholesky decomposition (https://github.com/pytorch/pytorch/issues/25785) or an eigendecomposition / SVD approach to compute the root decomposition (see https://github.com/pytorch/botorch/blob/3d8976f51d3bba0f730557ed61668d343df3ff1f/botorch/sampling/qmc.py#L136-L143 that actually references BoTorch code). The downside with that is that it will be much slower than just adding some small jitter.

**Comment by Balandat** [comment]
I'm working on a PR upstream in gpytorch to address this issue: https://github.com/cornellius-gp/gpytorch/pull/1105

**Comment by WeijingDai** [comment]
Hi all,
I also get stuck in a similar error 'cholesky_cpu: U(48,48) is zero, singular U'. In my case, the model has five variables and one output, and the model is FixedNoiseGP. If I want to keep the same initial starting points, any suggestion to investigate this problem or get across this error at least obtaining some result?
Thanks a lot
Weijing

**Comment by Balandat** [comment]
Can you describe more where in the optimization this is happening (both in terms of the stack trace and how many iterations you've run)? Often this will occur later on, when the optimization has already identified the location of the maximum, and all the test points are clustered together. In that case, a reasonable thing may be to just stop optimizing (if that is indicated by the"learning  curve" of the optimization).

**Comment by WeijingDai** [comment]
I working on an engineering optimization problems with 5 variables and 1 target. For each variable, 5 discrete choices are avaiable, meaning I would like to use BO to find the best combination of those variable choices. I followed the advice from you to build the loop on discrete domain. I do observe that the selection of Next Point was static or jumping around, but it was not the global maximum (I know where it is beforehand). I put the major part of the code here (I want the target variable close to zero).

       ` acqf_para = acqf_para_list[j]
        train_x = torch.as_tensor(train_data[:, 0:-1], dtype=dtype, device=device)
        train_y_origin = torch.as_tensor(train_data[:, -1], dtype=dtype, 
             device=device).unsqueeze(1)*1000`
        train_y = train_y_origin**2
        model = FixedNoiseGP(train_x, -train_y, train_Yvar=torch.full_like(train_y, 0.000001)).to(train_x)
        mll = ExactMarginalLogLikelihood(model.likelihood, model)
        fit_gpytorch_model(mll)

        # EI = ExpectedImprovement(model = model, best_f = -train_y.min() - acqf_para,)
        UCB = UpperConfidenceBound(model = model, beta = acqf_para)
        # PI = ProbabilityOfImprovement(model = model, best_f = -train_y.min() - acqf_para,)
        
        # Evaluate acquisition function over the discrete domain of parameters
        acqf = UCB(design_domain.unsqueeze(n_variables))
        acqf_flattened = acqf.flatten()
        acqf_sorted = torch.argsort(acqf.flatten(), descending=True)
        acqf_max = acqf_sorted[0].unsqueeze(0)
        for j in range(1, 10):
            if acqf_flattened[acqf_max[0]] > acqf_flattened[acqf_sorted[j]]:
                break
            else:
                acqf_max = torch.cat((acqf_max, acqf_sorted[j].unsqueeze(0)))

        candidate_id = acqf_max[torch.randint(len(acqf_max), size=(1, ))]
        candidate = design_domain_flattened[candidate_id]
        new_y = result[candidate_id]
        train_new_y_origin = torch.as_tensor([[new_y]], dtype=dtype, device=device)
        train_new_y = train_new_y_origin**2
        # update training points
        train_x = torch.cat([train_x, candidate])
        train_y = torch.cat([train_y, train_new_y])
        train_y_origin = torch.cat([train_y_origin, train_new_y_origin])`
       
I tried some different choices of beta for UCB, getting the same error. larger beta seems not necessarily proplong the iteration loop before the error happens. 

Btw, I also did a two-variable test under the same background setting, but the choices of the 2 variable are denser than the 5-variable problem (21 vs. 5). In this 2d problem, the same code also led to repeating choices but did not show the same error.

**Comment by Balandat** [comment]
Where exactly in the loop is the error happening? You're using a very small noise level (1e-6) on the `FixedNoiseGP`, I suspect this happens during training then?

In your observation data are there repeated observation at the same input locations? Are they on the same discrete grid, or is the input data on a continuous / different discrete domain? 

One thing you can do is run this on the `double` datatype (default is `float`). The improved precision can help working with badly conditioned covariance matrices. You can also increase the noise level some to see if that helps. I'd need a fully repro-able example to dig deeper.

**Comment by WeijingDai** [comment]
Thanks for your suggestion. 
1. The error happens durting the training, poping out from the FixedNoiseGP step. 
2. The grid are the same during the whole loop.
3. Some further investigation shows that it may be related to the scale of output. Using the same initial points, the error happens when using y**2, but nearly no error when using log(y**2 + 1e-16).
I guess that if the range of y is large (like more than 1000 max/min), the error easily appears.
4. It seems like the problem I meet is related to GP rathet than BO, and it give me the feeling that the scale of both input and ouput variable influence the GP.
5. The problem can be partially solved, but I run into new issue. Since the grid in my scenario is quite sparse (5 points at each dimension), the BO stop exploration quickly (when using UCB with beta smaller than 10). If the grid cannot be denser, any method I can improve this situation?

Here is my repo with the data and script if you would like to have a look.
https://github.com/WeijingDai/SiP_Optimization/tree/master/test

**Comment by Balandat** [comment]
> Some further investigation shows that it may be related to the scale of output

Yeah data normalization / standardization is quite important. Our default priors are geared towards using standardized (zero mean, unit variance) outcomes (and inputs/features normalized to the unit hypercube). 

> Since the grid in my scenario is quite sparse (5 points at each dimension), the BO stop exploration quickly

Not sure that's necessarily a bad thing, could just mean that the mean locally is high. You could try other acquisition functions that will ensure that you keep exploring, e.g. EI

**Comment by goodnamegood** [comment]
I'm also getting error like this, where i have hundreds of parameters to optimize, and I use train_X size of 10 x 128 where 128 is the dimension of my parameters. Please advise, thanks

**Comment by Balandat** [comment]
I'm closing this out for now - we do have some improvements for optimizing discrete parameters in the works so keep an eye out for them.
